### ISSUE TITLE
Curly rule option proposal: flow

### ISSUE THREAD

**Issue by arcanis** [eslint/eslint#5686]
**When does this rule warn?**

This proposal is about an additional option to the `curly` rule (I suggest to name it `flow`). Its goal is to help make reading the code extremely fluid.

  - An `if` or `for` statement should never use curly braces except in the following cases:

      - If it contains multiple statements (as always)
      - If it has an `else` counterpart that strongly needs curly braces *(weak)*
      - If it has no `else` counterpart, and contains a dedent token in its body
      - If it is the last statement of a block

  - An `else` statement should never use curly braces, except in the following cases:

      - If it contains multiple statements (as always)
      - If its `if` counterpart strongly needs curly braces *(weak)*
      - If it contains a dedent token in its body
      - If it is the last statement of a block

*Note about weak requirements:* Weak requirements are not accounted for in others weak requirements. Basically, it simply means that the two weak rules don't have any sort of circular dependency.

*Note about `else` staments:* When we say that an `else` statement "uses curly braces", it also take into account an `else` statement immediately followed by a `for`, `if`, `try`, `with`, or `switch` construct. In such a case, it means that this followup construct uses curly braces instead of the `else` statement itself.

*The following patterns are considered problems:*

```js
for (let result of results) {
    doSomething();
}

if (test)
    test = {
        foo: 42
    };

if (test)
    if (test2) {
        doSomething();
    }

if (test) {
    doSomething();
}

let x = function () {
    if (test)
        doSomething();
};

let y = function () {
    if (test)
        doSomething();
    else
        doSomethingElse();
};

if (test) {
    doSomething();
} else
    for (let result of results)
        doSomethingElse();
```

*The following patterns are **not** considered problems:*

```js
for (let result of results)
    doSomething();

if (test) {
    test = {
        foo: 42
    };
}

if (test)
    if (test2)
        doSomething();

if (test)
    doSomething();

let x = function () {
    if (test) {
        doSomething();
    }
};

let y = function () {
    if (test) {
        doSomething();
    } else {
        doSomethingElse();
    }
};

if (test) {
    doSomething();
} else for (let result of results) {
    doSomethingElse();
}

if (!results) {
    doSomething();
} else for (let result of results) {
    doSomethingElse(result);
}
```

**Is this rule preventing an error or is it stylistic?**

Stylistic.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

I firmly believe this rule allow for an easy-to-read lecture flow, and complements in a good way the already existing `curly` modes.

**Are you willing to create the rule yourself?**

I could help, but I might need some guidance.

**Comment by BYK** [comment]
This looks almost the same as `["multi-or-nest", "consistent"]` configuration. The only difference is the lack of "If it is the last statement of a block" part. I'm not sure if the additional name and complexity is worth for that one exception.

**Comment by arcanis** [comment]
I agree that it might be a bit too personal-preference to be in the ESLint core. I've implemented this rule in a [personal plugin](https://github.com/arcanis/eslint-plugin-arca), so if anyone needs it, it's there.

And actually, I have to say that I'm quite impressed by how easy it is to add custom rules to ESLint, tests and all. That's some really nice work.

**Comment by nzakas** [comment]
Agree with @BYK. Seems pretty close to what we already have, and since you already have a plugin that does this, I'd suggest continuing to use that.

Glad you found writing custom rules easy, that's the goal! :)
